### PR TITLE
Handle uploads errors

### PR DESCRIPTION
**Pull request by mikkio-j**
Issue: #474

### CODE REVIEW

File: src/hooks/useUploadsManager/useUploadsManager.tsx [hunk 1]
```diff
@@ -1,62 +1,187 @@
...
+    displaySnackbar({
+      title: `(${lostConnectionAssets.length}) Asset${
+        lostConnectionAssets.length > 1 ? 's' : ''
+      } waiting to resume upload`,
+      description: 'Reconnect files to fix the issue',
+      actionText: 'See assets',
+      onActionClick: () => navigate(absoluteRoutes.studio.uploads()),
+      iconType: 'warning',
+    })
+  }, [lostConnectionAssets.length])
```
Comment (drillprop) [inline src/hooks/useUploadsManager/useUploadsManager.tsx]: If we cannot do anything with it  let's just add above ` // eslint-disable-next-line react-hooks/exhaustive-deps`

File: src/hooks/useUploadsManager/useUploadsManager.tsx [hunk 2]
```diff
@@ -1,62 +1,187 @@
...
+    ) => {
       const setAssetUploadProgress = (progress: number) => {
         setUploadsProgress((prevState) => ({ ...prevState, [asset.contentId]: progress }))
       }
-
+      const isFileInState = assetsFiles?.find((file) => file.contentId === asset.contentId)
+      if (!isFileInState && file) {
+        setAssetsFiles([...assetsFiles, { contentId: asset.contentId, blob: file }])
+      }
+      const savedFile = assetsFiles.find((file) => file.contentId === asset.contentId)?.blob
```
Comment (drillprop) [inline src/hooks/useUploadsManager/useUploadsManager.tsx]: I don't think, it's necessary. You used `find` on line 116, right?

File: src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx [hunk 3]
```diff
@@ -12,26 +17,116 @@ import {
...
+          {
+            contentId: asset.contentId,
+            owner: asset.owner,
+            parentObject: {
+              type: asset.parentObject.type,
+              id: asset.parentObject.id,
+            },
+            type: asset.type,
+          },
+          randomStorageProviderUrl ?? '',
```
Comment (drillprop) [inline src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx]: not sure, but I think we shouldn't run `startFileUpload` if the `randomStorageProviderUrl` is `falsy`.

File: src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx [hunk 4]
```diff
@@ -12,26 +17,116 @@ import {
...
+      {
+        contentId: asset.contentId,
+        owner: asset.owner,
+        parentObject: {
+          type: asset.parentObject.type,
+          id: asset.parentObject.id,
+        },
+        type: asset.type,
+      },
+      randomStorageProviderUrl ?? '',
```
Comment (drillprop) [inline src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx]: Same as above

File: src/hooks/useUploadsManager/types.ts [hunk 5]
```diff
@@ -28,9 +28,20 @@ export type InputAssetUpload = Omit<AssetUpload, 'lastStatus' | 'size'>
...
 
+export type StartFileUploadOptions = {
+  isReUpload?: boolean
+  changeHost?: boolean
+}
+
 export type UploadManagerValue = {
-  uploadsState: AssetUploadWithProgress[]
-  startFileUpload: (file: File | Blob, asset: InputAssetUpload, storageMetadata: string) => void
+  uploadsState: AssetUploadWithProgress[][]
```
Comment (kdembler) [inline src/hooks/useUploadsManager/types.ts:42]: Why is this a nested array?

File: src/hooks/useUploadsManager/useUploadsManager.tsx [hunk 6]
```diff
@@ -1,62 +1,188 @@
...
+    progress: uploadsProgress[asset.contentId] ?? 0,
+  }))
+
+  const channelDataObjects = [channel?.avatarPhotoDataObject, channel?.coverPhotoDataObject]
+  const videosDataObjects = videos?.flatMap((video) => [video.mediaDataObject, video.thumbnailPhotoDataObject]) || []
+  const allDataObjects = [...channelDataObjects, ...videosDataObjects]
+
+  // Enriching data with pending/accepted/rejected status
+  const uploadsStateWithLiaisonJudgement = uploadsStateWithProgress.map((asset) => {
+    const dataObject = allDataObjects.find((dataObject) => dataObject?.joystreamContentId === asset.contentId)
```
Comment (kdembler) [inline src/hooks/useUploadsManager/useUploadsManager.tsx]: Maybe we could add a warning if `dataObject` is not found? That would point to some local state corruption, might be useful for development

File: src/hooks/useUploadsManager/useUploadsManager.tsx [hunk 7]
```diff
@@ -1,62 +1,188 @@
...
+    displaySnackbar({
+      title: `(${lostConnectionAssets.length}) Asset${
+        lostConnectionAssets.length > 1 ? 's' : ''
+      } waiting to resume upload`,
+      description: 'Reconnect files to fix the issue',
+      actionText: 'See assets',
+      onActionClick: () => navigate(absoluteRoutes.studio.uploads()),
+      iconType: 'warning',
+    })
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (kdembler) [inline src/hooks/useUploadsManager/useUploadsManager.tsx]: Why do we need that, what dependency do we want to ignore?

File: src/hooks/useUploadsManager/useUploadsManager.tsx (hunk 6 above)
Comment (kdembler) [reply src/hooks/useUploadsManager/useUploadsManager.tsx]: I guess we would display it only if we're not loading data anymore

File: src/hooks/useUploadsManager/useUploadsManager.tsx [hunk 8]
```diff
@@ -1,62 +1,188 @@
...
+  }, [lostConnectionAssets.length])
+
+  // Enriching video type assets with video title
+  const uploadsStateWithVideoTitles = uploadsStateWithLiaisonJudgement.map((asset) => {
+    if (asset.type === 'video') {
+      const video = videos?.find((video) => video.mediaDataObject?.joystreamContentId === asset.contentId)
```
Comment (kdembler) [inline src/hooks/useUploadsManager/useUploadsManager.tsx:99]: Likewise, we could show a warning here

File: src/hooks/useUploadsManager/useUploadsManager.tsx [hunk 9]
```diff
@@ -1,62 +1,188 @@
...
+      const key = asset.parentObject.id
+      !acc[key] ? (acc[key] = [{ ...asset }]) : acc[key].push(asset)
+      return acc
+    }, {})
+  )
 
   const startFileUpload = useCallback(
-    async (file: File | Blob, asset: InputAssetUpload, storageMetadata: string) => {
+    async (
+      file: File | Blob | null,
```
Comment (kdembler) [inline src/hooks/useUploadsManager/useUploadsManager.tsx:130]: We need to do one of the following:
1) Handle case where both `file` and `fileInState` are null
2) Create a separate fn for resuming/retrying a download

File: src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx [hunk 10]
```diff
@@ -12,26 +17,122 @@ import {
...
+          {
+            isReUpload: true,
+          }
+        )
+      }
+    },
+    [asset, randomStorageProviderUrl, startFileUpload]
+  )
+
+  const { getRootProps, getInputProps, open } = useDropzone({
```
Comment (kdembler) [inline src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx]: Let's rename `open` on destructure so it's more descriptive

File: src/hooks/useUploadsManager/types.ts (hunk 5 above)
Comment (kdembler) [reply src/hooks/useUploadsManager/types.ts:42]: Is this about grouping the assets together? So something like?
```
[
  [video-1-thumbnail, video-1-media],
  [video-2-thumbnail, video-2-media],
]
```

**Review by kdembler — COMMENTED** [review]
Just the code for now, I'll test it out later because the query node is down

File: src/hooks/useUploadsManager/types.ts (hunk 5 above)
Comment (mikkio-j) [reply src/hooks/useUploadsManager/types.ts:42]: > Is this about grouping the assets together? So something like?
> 
> ```
> [
>   [video-1-thumbnail, video-1-media],
>   [video-2-thumbnail, video-2-media],
> ]
> ```

Yes, that's exactly it

File: src/hooks/useUploadsManager/useUploadsManager.tsx (hunk 7 above)
Comment (mikkio-j) [reply src/hooks/useUploadsManager/useUploadsManager.tsx]: `displaySnackbar` is causing updates. Wasn't sure how should I handle that

Comment (kdembler) [reply src/hooks/useUploadsManager/useUploadsManager.tsx]: Almost always when we disable `react-hooks/exhaustive-deps` something will break. Not including `displaySnackbar` in the dependency array will cause the hook to always use the stale reference to the function. `displaySnackbar` in turn uses `...snackbars` so using old `displaySnackbar` reference may cause some of the snackbars to disappear.

Disabling this rule is almost always not the right solution. In this case, we should just make the `displaySnackbar` reference stable

File: src/hooks/useUploadsManager/useUploadsManager.tsx [hunk 11]
```diff
@@ -1,62 +1,201 @@
...
+    })
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [lostConnectionAssets.length, navigate])
+
+  // Enriching video type assets with video title
+  const uploadsStateWithVideoTitles = uploadsStateWithLiaisonJudgement.map((asset) => {
+    if (asset.type === 'video') {
+      const video = videos?.find((video) => video.mediaDataObject?.joystreamContentId === asset.contentId)
+      if (!video && !videosLoading) {
+        console.warn(`Video not found. ContentId: ${asset.contentId}`)
```
Comment (kdembler) [inline src/hooks/useUploadsManager/useUploadsManager.tsx]: When doing upload I get a lot (like 30 or something) warning like this in the console. Could you look into that?

File: src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx [hunk 12]
```diff
@@ -47,19 +148,43 @@ const AssetLine: React.FC<AssetLineProps> = ({ isLast = false, asset }) => {
...
+        showDialog={showDialog}
+        variant="warning"
+        onSecondaryButtonClick={() => {
+          setShowDialog(false)
+          if (asset.parentObject.type === 'video') {
+            navigate(absoluteRoutes.studio.editVideo())
+          }
+          if (asset.parentObject.type === 'channel') {
+            navigate(absoluteRoutes.studio.editChannel())
+          }
```
Comment (kdembler) [inline src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx:164]: I think we got a misunderstanding on what's supposed to happen here. You redirect the user to "new video" screen when they click `Upload new`. But what should actually happen is that we create an update transaction with the new file and try to submit it, right from this screen and then start the upload. That's also what's in the diagram linked in the issue. However, maybe let's try to elevate that to separate task so we don't drag this PR for too long

File: src/hooks/useUploadsManager/useUploadsManager.tsx [hunk 13]
```diff
@@ -1,62 +1,212 @@
...
+    if (asset?.type === 'video') {
+      const video = videos?.find((video) => video.mediaDataObject?.joystreamContentId === asset.contentId)
+      const title = video?.title ?? null
+      return { ...asset, title }
+    }
+    return asset
+  })
+
+  // Check if liaison data and video title is available
+  uploadsStateWithVideoTitles.map((asset) => {
```
Comment (kdembler) [inline src/hooks/useUploadsManager/useUploadsManager.tsx:107]: In general, in cases like this where you don't care about return value but just want to iterate over the values, `.forEach` may be a better choice. It's a bit more performant, although negligibly, but mostly just for readability improvement

No need to change here

File: src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx [hunk 14]
```diff
@@ -11,28 +16,122 @@ import {
...
 
-  const renderStatusIndicator = (asset: UploadData) => {
+  const renderStatusMessage = (asset: AssetUploadWithProgress) => {
+    if (asset.lastStatus === 'reconnecting') {
+      return 'Trying to reconnect...'
+    }
+    if (asset.lastStatus === 'reconnectionError') {
+      return (
+        <Button size="small" variant="secondary" icon={<SvgGlyphUpload />} onClick={handleChangeHost}>
+          Change host (0.2 JOY)
```
Comment (kdembler) [inline src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx]:
```suggestion
          Change host
```

File: src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx [hunk 15]
```diff
@@ -47,19 +146,43 @@ const AssetLine: React.FC<AssetLineProps> = ({ isLast = false, asset }) => {
...
-        </FileInfoType>
-        <Text variant="body2">{dimension}</Text>
-        <Text>{size}</Text>
-      </FileInfoContainer>
-      <StatusMessage variant="subtitle2">{fileStatusMessage}</StatusMessage>
-    </FileLineContainer>
+    <>
+      <MessageDialog
+        title="Different file was selected!"
+        description="We detected that you selected a different file than the one you uploaded previously. Select the same file to continue the upload or make a new transaction (0.2JOY) and upload a new version of file."
```
Comment (kdembler) [inline src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx]: I'd like the copy here to be `We detected that you selected a different file than the one you uploaded previously. Select the same file to continue the upload or edit "your channel"/"the video" to use the new file.`

We should show the correct variant depending on whether it's a video or a channel

File: src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx [hunk 16]
```diff
@@ -47,19 +146,43 @@ const AssetLine: React.FC<AssetLineProps> = ({ isLast = false, asset }) => {
...
+          if (asset.parentObject.type === 'channel') {
+            navigate(absoluteRoutes.studio.editChannel())
+          }
+        }}
+        primaryButtonText="Reselect file"
+        onPrimaryButtonClick={() => {
+          setShowDialog(false)
+          openFileSelect()
+        }}
+        secondaryButtonText="Upload new (0.2 JOY)"
```
Comment (kdembler) [inline src/views/studio/MyUploadsView/AssetLine/AssetLine.tsx]: `Edit channel` / `Edit video` same as above

**Review by kdembler — APPROVED** [review]

**Pull request merged**